core/drivers/pl022_spi.c

```diff
@@ -158,155 +157,82 @@
 #define SSP_SCR_MIN		0
 #define SSP_DATASIZE_MAX	16
 
-enum pl022_data_size {
-	PL022_DATA_SIZE4 = 0x3,
-	PL022_DATA_SIZE5,
-	PL022_DATA_SIZE6,
-	PL022_DATA_SIZE7,
-	PL022_DATA_SIZE8 = SSPCR0_DSS_8BIT,
-	PL022_DATA_SIZE9,
-	PL022_DATA_SIZE10,
-	PL022_DATA_SIZE11,
-	PL022_DATA_SIZE12,
-	PL022_DATA_SIZE13,
-	PL022_DATA_SIZE14,
-	PL022_DATA_SIZE15,
-	PL022_DATA_SIZE16 = SSPCR0_DSS_16BIT
-};
-
-enum pl022_spi_mode {
-	PL022_SPI_MODE0 = SSPCR0_SPO0 | SSPCR0_SPH0, /* 0x00 */
-	PL022_SPI_MODE1 = SSPCR0_SPO0 | SSPCR0_SPH1, /* 0x80 */
-	PL022_SPI_MODE2 = SSPCR0_SPO1 | SSPCR0_SPH0, /* 0x40 */
-	PL022_SPI_MODE3 = SSPCR0_SPO1 | SSPCR0_SPH1  /* 0xC0 */
-};
-
-static void pl022_txrx8(struct spi_chip *chip, uint8_t *wdat,
-	uint8_t *rdat, size_t num_txpkts, size_t *num_rxpkts)
+static enum spi_result pl022_txrx8(struct spi_chip *chip, uint8_t *wdat,
+	uint8_t *rdat, size_t num_pkts)
 {
 	size_t i = 0;
 	size_t j = 0;
 	struct pl022_data *pd = container_of(chip, struct pl022_data, chip);
 
-	pd->gpio->ops->set_value(pd->cs_gpio_pin, GPIO_LEVEL_LOW);
-
-	while (i < num_txpkts) {
-		if (read8(pd->base + SSPSR) & SSPSR_TNF)
-			/* tx 1 packet */
-			write8(wdat[i++], pd->base + SSPDR);
+	if (pd->data_size_bits != 8) {
+		EMSG("data_size_bits should be 8, not %u\n",
+			pd->data_size_bits);
+		return SPI_ERR_CFG;
 	}
 
-	do {
-		while ((read8(pd->base + SSPSR) & SSPSR_RNE) &&
-			(j < *num_rxpkts))
-			/* rx 1 packet */
-			rdat[j++] = read8(pd->base + SSPDR);
-	} while ((read8(pd->base + SSPSR) & SSPSR_BSY) && (j < *num_rxpkts));
-
-	*num_rxpkts = j;
-
-	pd->gpio->ops->set_value(pd->cs_gpio_pin, GPIO_LEVEL_HIGH);
-}
-
-static void pl022_txrx16(struct spi_chip *chip, uint16_t *wdat,
-	uint16_t *rdat, size_t num_txpkts, size_t *num_rxpkts)
-{
-	size_t i = 0;
-	size_t j = 0;
-	struct pl022_data *pd = container_of(chip, struct pl022_data, chip);
-
-	pd->gpio->ops->set_value(pd->cs_gpio_pin, GPIO_LEVEL_LOW);
+	if (wdat)
```

Comment:
A comment that is out of scope of the PR: the SPI IP is likely to rely on some internal FIFO buffer of limited size. If one requests a exchange of more byte than the IP FIFO size, the IP will fail to buffer the full received SPI frame while data are sent, leading to some FIFO overflow.

Another one: if `rdat` is `NULL`, FIFO will not be wiped. Possible FIFO overflow issue. How will the IP behave ?
And another one,related: if `wdat` is NULL, IP won't transmit data: will the SPI communication operate in such condition? I think your test should treat those cases: only one of the `rdat` and `wdat` is `NULL`).

Am I right ? Should we open a Issue for that ? (this issue is out of scope of the PR, as former implementation used the same 'transmit all then receive all' scheme)